### PR TITLE
Measure peak memory usage for stage tasks execution.

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java [hunk 1]
```diff
@@ -390,6 +360,51 @@ public String toString()
...
+    private class StageTaskListener
+            implements StateChangeListener<TaskInfo>
+    {
+        private long previousMemory;
+
+        @Override
+        public void stateChanged(TaskInfo taskInfo)
+        {
+            long currentMemory = taskInfo.getStats().getMemoryReservation().toBytes();
+            long deltaMemoryInBytes = currentMemory - previousMemory;
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java]: This needs to be made thread safe. `stateChanged` is invoked from different threads in `StateMachine.fireStateChanged`, possibly even concurrently

File: presto-main/src/main/java/com/facebook/presto/execution/StageStateMachine.java [hunk 2]
```diff
@@ -182,6 +199,7 @@ public StageInfo getStageInfo(Supplier<Iterable<TaskInfo>> taskInfosSupplier, Su
         int completedDrivers = 0;
 
         long totalMemoryReservation = 0;
+        long peekMemoryReservation = getPeakMemoryInBytes();
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/execution/StageStateMachine.java]: typo, `peak` not `peek`

File: presto-main/src/main/java/com/facebook/presto/execution/StageStats.java [hunk 3]
```diff
@@ -49,6 +49,7 @@
     private final int completedDrivers;
 
     private final DataSize totalMemoryReservation;
+    private final DataSize peekMemoryReservation;
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/execution/StageStats.java]: same typo

**Comment by cberner** [comment]
Sorry it took me so long to get to this. Had some comments

File: presto-main/src/main/java/com/facebook/presto/execution/StageStats.java (hunk 3 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/execution/StageStats.java]: done

File: presto-main/src/main/java/com/facebook/presto/execution/StageStateMachine.java (hunk 2 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/execution/StageStateMachine.java]: done

File: presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java (hunk 1 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java]: Can it be called concurrently? This is exactly the same mechanism that is being used in existing `MemoryTrackingRemoteTaskFactory.UpdatePeakMemory` class. I though that only `updateMemoryUsage` should be thread safe. Could it be possible that also task peak memory tracking is broken? I can fix both of them by using `AtomicLong`

**Comment by sopel39** [comment]
rebased and applied comments FYI: @cberner

File: presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java (hunk 1 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java]: The difference is that `UpdatePeakMemory.stateChanged` is `synchronized`

Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java]: Right, thanks, somehow I didn't notice that.

Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java]: fixed

**Comment by cberner** [comment]
Looks good! We're doing a release soon, so will merge after that's done

**Comment by cberner** [comment]
Merged. Thanks!

**Pull request closed without merging**